File: Shared/Models/Challenge/ChallengeContent.swift
//
//  ChallengeContent.swift
//  CKRAdmin
//
//  Created by Julien Rahier on 11/05/2025.
//

import Foundation

enum ChallengeContent: Equatable, Codable, Hashable {
    case picture(PictureContent)
    case multipleChoice(MultipleChoiceContent)
    case singleAnswer(SingleAnswerContent)
    case noChoice(NoChoiceContent)
}

extension ChallengeContent {
    var toResponseContent: ChallengeResponseContent {
        switch self {
            case .picture:
                return .picture("")
            case .multipleChoice:
                return .multipleChoice([])
            case .singleAnswer:
                return .singleAnswer("")
            case .noChoice:
                return .noChoice
        }
    }
}

File: android/app/src/main/java/dev/rahier/colocskitchenrace/data/model/ChallengeContent.kt
package dev.rahier.colocskitchenrace.data.model

sealed class ChallengeContent {
    data object Picture : ChallengeContent()
    data class MultipleChoice(
        val choices: List<String> = listOf("", "", "", ""),
        val correctAnswerIndex: Int? = null,
        val shuffleAnswers: Boolean = true,
    ) : ChallengeContent()

    data object SingleAnswer : ChallengeContent()
    data class NoChoice(val text: String = "") : ChallengeContent()

    val type: String
        get() = when (this) {
            is Picture -> "picture"
            is MultipleChoice -> "multipleChoice"
            is SingleAnswer -> "singleAnswer"
            is NoChoice -> "noChoice"
        }

    fun toResponseContent(): ChallengeResponseContent = when (this) {
        is Picture -> ChallengeResponseContent.Picture("")
        is MultipleChoice -> ChallengeResponseContent.MultipleChoice(emptyList())
        is SingleAnswer -> ChallengeResponseContent.SingleAnswer("")
        is NoChoice -> ChallengeResponseContent.NoChoice
    }
}
